### DISCUSSION TITLE
Python Package Help

### DISCUSSION THREAD

**Discussion by zos474** [rhasspy/piper#326] [Q&A] [answered]
Hello all, appreciate some guidance here. I'm trying to run the piper package in Python and there were some tips from @rhohndorf at https://github.com/rhasspy/piper/discussions/215#discussioncomment-7063021. By trial and error, I manage to find out how to import the package (i.e. import piper.voice, rather than import pipertts, etc.) and have gotten as far as loading a model, but now stuck at the various parameters needed for voice.synthesize. Specifically, not clear to me how to define/set up the wav file needed. Is there a help file or some simple demo code I can use as a reference?
~~~import os
from piper.voice import PiperVoice

voicedir = os.path.expanduser('~/Documents/piper/')
model = voicedir+"en_GB-alan-medium.onnx"
voice = PiperVoice.load(model)

import wave
#How to set up the wav file here?

text = "This is an example of text-to-speech using Piper TTS."
audio = voice.synthesize(text,???)

**Comment by zos474** [comment, marked as answer]
OK, thank goodness for codeium, I managed to figure it out and leaving this as a starter for others as new to this as I am...

~~~
import os
import wave
from piper.voice import PiperVoice

voicedir = os.path.expanduser('~/Documents/piper/') #Where onnx model files are stored on my machine
model = voicedir+"en_GB-alan-medium.onnx"
voice = PiperVoice.load(model)
wav_file = wave.open('output.wav', 'w')
text = "This is an example of text-to-speech using Piper TTS."
audio = voice.synthesize(text,wav_file)

**Reply by slowrunner** [reply]
Super! Thanks for posting this  "piper2wav.py" 

Now if I can figure out to output to direct to my audio device without a temp file in the middle, I'll be set.

This works on my Raspberry Pi5 ... but creates a temp.wav file and then plays it:
```
#!/bin/env python3

# REQ:
#  sudo pip3 install sounddevice --break-system-packages
#  sudo pip3 install soundfile --break-system-packages
#  sudo pip3 install piper-tts --break-system-packages


import sounddevice as sd
import soundfile as sf
import wave
import os
from piper.voice import PiperVoice


filename = 'temp.wav'

voicedir = os.path.expanduser('~/wali_pi5/c3ws/models/piper-tts/') #Where onnx model files are stored on my machine
model = voicedir+"en_US-arctic-medium.onnx"
voice = PiperVoice.load(model)
wav_file = wave.open(filename, 'w')
text = "This temporary wave file was generated from Python call to piper output to python sound device"
audio = voice.synthesize(text,wav_file)

# Extract data and sampling rate from file
data, fs = sf.read(filename, dtype='float32')
sd.play(data, fs)
status = sd.wait()  # Wait until file is done playing
os.remove(filename)

```

**Reply by JoseEliel** [reply]
Hi! Did you manage to figure out how to avoid saving the file?
If not, I found how to do it by reading the methods within piper-tts. Here is how I do it:

```
import sounddevice as sd
from piper.voice import PiperVoice

voicedir = "./voices/" #Where onnx model files are stored on my machine
model = voicedir+"en_US-lessac-medium.onnx"
voice = PiperVoice.load(model)
text = "This is an example of text-to-speech using Piper TTS."

# Setup a sounddevice OutputStream with appropriate parameters
# The sample rate and channels should match the properties of the PCM data
stream = sd.OutputStream(samplerate=voice.config.sample_rate, channels=1, dtype='int16')
stream.start()

for audio_bytes in voice.synthesize_stream_raw(text):
    int_data = np.frombuffer(audio_bytes, dtype=np.int16)
    stream.write(int_data)

stream.stop()
stream.close()
```

**Reply by mitchind** [reply]
Appreciate your examples. I've been getting write errors using piper examples to write the file and stream it.
In your example it looks like it's missing np definition
I get 
on the line `int_data = np.frombuffer(audio_bytes, dtype=np.int16)`
> NameError: name 'np' is not `defined`

**Reply by mitchind** [reply]
Got it - `import numpy as np`

Still getting same issue that audio is cut off in mid play - looks like script is stuck.
Cutoff audio is same symptom when writing to a file.

**Reply by mitchind** [reply]
And cutoff audio fixed - was a usb speaker that didn't have enough power from Raspberry Pi USB - had to run through a powered hub
